File: delegatervadapter/src/main/java/com/xiansenliu/delegatervadapter/DelegateAdapter.java
package com.xiansenliu.delegatervadapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;


import java.util.List;

/**
 * Created by xinliu
 * Date       17-3-4
 * Time       13:17.
 */

public class DelegateAdapter<T> extends RecyclerView.Adapter {
    private static final String TAG = "DelegateAdapter";
    private List<T> mItems;

    private DelegateManager mDelegateManager;

    public DelegateAdapter(List<T> items) {
        this(items, new DefaultManager());
    }

    public DelegateAdapter(List<T> items, DelegateManager delegateManager) {
        this.mItems = items;
        mDelegateManager = new DefaultManager();
    }

    /**
     * Add delegate.
     *
     * @param delegates at least one delegate
     */
    public void addDelegates(@NonNull ViewDelegate... delegates) {
        if (delegates.length == 0) {
            try {
                throw new Exception("delegates' size should never be zero ");
            } catch (Exception e) {
                e.printStackTrace();
            }
            return;
        }
        for (ViewDelegate delegate : delegates) {
            mDelegateManager.addDelegate(delegate);
        }
    }

    public void removeDelegate(@NonNull ViewDelegate delegates) {
        mDelegateManager.removeDelegate(delegates);
    }

    @Override
    public int getItemCount() {

        return mItems.size();
    }

    @Override
    public int getItemViewType(int position) {
        int viewType = mDelegateManager.getItemViewType(mItems, position);

        return viewType;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder vh = mDelegateManager.onCreateViewHolder(parent, viewType);
        return vh;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        mDelegateManager.onBindViewHolder(holder, mItems, position, null);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position, List payloads) {
        mDelegateManager.onBindViewHolder(holder, mItems, position, payloads);
    }
}

File: delegatervadapter/src/main/kotlin/com/xiansenliu/delegatervadapter/DelegateAdapter.kt
package com.xiansenliu.delegatervadapter

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

/**
 * Author       xinliu
 * Date         8/28/17
 * Time         4:43 PM
 */
class DelegateAdapter(items: List<Any>,
                      private val manager: DelegateManager = DelegateManager())
    : RecyclerView.Adapter<BaseVH<Any>>() {
    private var items: MutableList<Any> = mutableListOf()

    init {
        this.items.addAll(items)
    }

    fun addHolderDelegate(vararg delegates: ModelDelegate<*>) {
        manager.addDelegates(*delegates)
    }

    override fun getItemCount(): Int {
        return items.size
    }


    override fun getItemViewType(position: Int): Int {
        val type = manager.getItemViewType(position, items)
        return type
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): BaseVH<Any> {
        val vh = manager.onCreateViewHolder(parent, viewType, items)
        return vh
    }

    override fun onBindViewHolder(holder: BaseVH<Any>?, position: Int) {
        manager.onBindViewHolder(holder, position, items)
    }


    fun updateItems(items: List<Any>) {
        with(this.items) {
            clear()
            addAll(items)
        }
        notifyDataSetChanged()
    }

    fun insertItems(items: List<Any>, index: Int = this.items.size) {
        this.items.addAll(index, items)
        notifyItemRangeInserted(index, items.size)
    }

    fun removeItems(items: List<Any>) {
        this.items.removeAll(items)
        notifyDataSetChanged()
    }

}